File: Android/App/app/src/main/java/com/efalco777/crud_products/presentation/product/viewmodel/products/ProductsViewModel.kt
package com.efalco777.crud_products.presentation.product.viewmodel.products

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.efalco777.crud_products.domain.product.use_case.GetProductsUseCase
import com.efalco777.crud_products.presentation.core.viewmodel.DataState
import com.efalco777.crud_products.presentation.product.mapper.DomainToPresentationProductMapper
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import javax.inject.Inject

@HiltViewModel
class ProductsViewModel @Inject constructor(
    private val getProductsUseCase: GetProductsUseCase
) : ViewModel() {
    private val _state = MutableStateFlow(ProductsViewModelState.initial())
    val flow = _state.asStateFlow()

    init {
        loadData()
    }

    fun loadData() {
        try {
            _state.update { currentState ->
                currentState.copy(dataState = DataState.Loading)
            }
            getProductsUseCase.invoke(Unit, viewModelScope, onResult = { result ->
                _state.update { _ ->
                    ProductsViewModelState(
                        products = result.map { model -> DomainToPresentationProductMapper.map(model) },
                        dataState = DataState.Loaded,
                    )
                }
            })

        } catch (e: Exception) {
            e.printStackTrace()
            _state.update { _ ->
                _state.value.copy(dataState = DataState.Error(e))
            }
        }
    }
}

File: iOS/App/ProductsCrud/Presentation/Product/ViewModel/Products/ProductsViewModel.swift
//
//  ProductsViewModel.swift
//  ProductsCrud
//
//  Created by Emil Falkowski on 29/10/2024.
//

import Combine
import OSLog
import SwiftUI

class ProductsViewModel: ObservableObject {
    private static let logger = Logger.create(for: ProductsViewModel.self)

    private let useCase: GetProductsUseCase

    @Published var state: ProductsViewModelState =
        ProductsViewModelState.initial()

    init(useCase: GetProductsUseCase) {
        self.useCase = useCase
    }

    @MainActor
    func fetchProducts() async {
        do {
            let products = try await useCase.execute()
            state = ProductsViewModelState(
                dataState: DataState.loaded,
                products: products.map {
                    DomainToPresentationProductMapper.map(model: $0)
                })
        } catch {
            Self.logger.error("Error: \(error.localizedDescription)")
            state = ProductsViewModelState(
                dataState: DataState.error(error), products: [])
        }
    }
}
